File: python/pedigree.py
# coding: utf-8
# pedigree.py

from __future__ import annotations
from typing import Iterator, Optional
import sys

import error_codes
from exception_with_code import *
from common import read_csv


#################### FormatException ####################

class FormatException(ExceptionWithCode):
	def __init__(self, lines: list[str]):
		super().__init__(FormatException.create_message(lines))
	
	def get_error_code(self) -> error_codes.Type:
		return error_codes.Type.PEDIGREE_INVALID_FORMAT
	
	@staticmethod
	def create_message(lines: list[str]) -> str:
		if len(lines) == 1:
			s = "error : the following line doesn't have four columns :"
		else:
			s = "error : the following lines don't have four columns :"
		
		for line in lines:
			s += '\n' + line
		return s


#################### ParentsException ####################

class ParentsException(ExceptionWithCode):
	def __init__(self, lines: list[str]):
		super().__init__(ParentsException.create_message(lines))
	
	def get_error_code(self) -> error_codes.Type:
		return error_codes.Type.PARENT_NOT_DEFINED
	
	@staticmethod
	def create_message(parents: list[str]) -> str:
		if len(parents) == 1:
			s = "error : the following parent isn't defined :"
		else:
			s = "error : the following parents aren't defined :"
		
		for parent in parents:
			s += '\n' + parent
		return s


#################### SamplesException ####################

class SamplesException(ExceptionWithCode):
	def __init__(self, lines: list[str]):
		super().__init__(SamplesException.create_message(lines))
	
	def get_error_code(self) -> error_codes.Type:
		return error_codes.Type.SAMPLES_NOT_IN_PEDIGREE
	
	@staticmethod
	def create_message(samples: list[str]) -> str:
		if len(samples) == 1:
			s = "error : the following sample isn't in pedigree :"
		else:
			s = "error : the following samples aren't in pedigree :"
		
		for sample in samples:
			s += '\n' + sample
		return s


#################### Family ####################

class Family:
	def __init__(self, mat: str, pat: str, progs: list[str]):
		self.mat: str				= mat
		self.pat: str				= pat
		self.progenies: list[str]	= progs
	
	def parents(self) -> tuple[str, str]:
		return (self.mat, self.pat)
	
	def num_progenies(self) -> int:
		return len(self.progenies)
	
	def samples(self) -> list[str]:
		return [self.mat, self.pat] + self.progenies


#################### Progeny ####################

class Progeny:
	def __init__(self, v: list[str]):
		self.family: str	= v[0]
		self.name: str		= v[1]
		self.mat: str		= v[2]
		self.pat: str		= v[3]
	
	def parents(self) -> tuple[str, str]:
		return (self.mat, self.pat)


#################### PedigreeTable ####################

class PedigreeTable:
	def __init__(self, progs: list[Progeny]):
		self.table: list[Progeny]	= progs
	
	def __len__(self) -> int:
		return len(self.table)
	
	def filter_with_parents(self, samples: list[str]) -> PedigreeTable:
		progs = [ p for p in self.table
						if p.mat != '0' and p.pat != '0' and
							p.mat in samples and p.pat in samples ]
		return PedigreeTable(progs)
	
	def generate_parents(self) -> Iterator[tuple[str, str]]:
		return (p.parents() for p in self.table)
	
	def get_children(self, parents: tuple[str, str]) -> list[str]:
		return [ p.name for p in self.table if p.parents() == parents ]
	
	def extract_families(self) -> list[Family]:
		parents = sorted(set((prog.mat, prog.pat) for prog in self.table))
		return [ Family(*p, self.get_children(p)) for p in parents ]
	
	def limit_samples(self, samples: list[str]) -> PedigreeTable:
		ped_samples = set(p.name for p in self.table)
		missing_samples = [ s for s in samples if s not in ped_samples ]
		if missing_samples:
			raise SamplesException(missing_samples)
		
		set_samples = set(samples)
		
		progs = [ p for p in self.table if p.name in set_samples ]
		return PedigreeTable(progs)
	
	def check_parents(self) -> list[str]:
		missing_parents = set()
		set_progs = set(prog.name for prog in self.table)
		for prog in self.table:
			for p in prog.parents():
				if p != '0' and p not in set_progs:
					missing_parents.add(p)
		return list(missing_parents)
	
	@staticmethod
	def read_lines(path: str) -> list[list[str]]:
		table: list[list[str]] = []
		not_four_columns_lines: list[str] = []
		try:
			with open(path, 'r') as f:
				for line in f:
					v = line.split()
					if len(v) != 4:
						not_four_columns_lines.append(line.rstrip())
					else:
						table.append(v)
			if not_four_columns_lines:
				raise FormatException(not_four_columns_lines)
		except FileNotFoundError:
			raise FileNotFoundException(path)
		except IOError:
			raise FileNotFoundException(path)
		
		return table
	
	@staticmethod
	def remove_duplidated_progenies(progs: list[Progeny]) -> list[Progeny]:
		new_progs: list[Progeny] = []
		s = set()
		for prog in progs:
			if prog.name not in s:
				new_progs.append(prog)
				s.add(prog.name)
			else:
				print('Progeny %s is duplicated.' % prog.name)
		return new_progs
	
	@staticmethod
	def create(progs: list[Progeny]) -> PedigreeTable:
		ped = PedigreeTable(progs)
		missing_parents = ped.check_parents()
		if missing_parents:
			raise ParentsException(missing_parents)
		return ped
	
	@staticmethod
	def read(path: str) -> PedigreeTable:
		table = PedigreeTable.read_lines(path)
		progs: list[Progeny] = [ Progeny(v) for v in table ]
		new_progs = PedigreeTable.remove_duplidated_progenies(progs)
		return PedigreeTable.create(new_progs)

File: src/Pedigree.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <cassert>
#include "../include/common.h"
#include "../include/Pedigree.h"

using namespace std;


//////////////////// FormatException ////////////////////

FormatException::FormatException(const vector<string>& lines) {
	stringstream	ss;
	if(lines.size() == 1)
		ss << "error : the following line doesn't have four columns :";
	else
		ss << "error : the following lines don't have four columns :";
	
	for(auto p = lines.begin(); p != lines.end(); ++p)
		ss << '\n' << *p;
	
	message = ss.str();
}

const char *FormatException::what() const noexcept {
	return message.c_str();
}


//////////////////// ParentsException ////////////////////

ParentsException::ParentsException(const vector<string>& parents) {
	stringstream	ss;
	if(parents.size() == 1)
		ss << "error : the following parent isn't defined :";
	else
		ss << "error : the following parents aren't defined :";
	
	for(auto p = parents.begin(); p != parents.end(); ++p)
		ss << '\n' << *p;
	
	message = ss.str();
}

const char *ParentsException::what() const noexcept {
	return message.c_str();
}


//////////////////// SamplesException ////////////////////

SamplesException::SamplesException(const vector<string>& samples) {
	stringstream	ss;
	if(samples.size() == 1)
		ss << "error : the following sample isn't in pedigree :";
	else
		ss << "error : the following samples aren't in pedigree :";
	
	for(auto p = samples.begin(); p != samples.end(); ++p)
		ss << '\n' << *p;
	
	message = ss.str();
}

const char *SamplesException::what() const noexcept {
	return message.c_str();
}


//////////////////// Family ////////////////////

Family::~Family() {
	for(auto p = progenies.begin(); p != progenies.end(); ++p)
		delete *p;
}

vector<string> Family::collect_samples() const {
	vector<string>	samples;
	samples.push_back(this->mat);
	samples.push_back(this->pat);
	for(auto p = this->progenies.begin(); p != this->progenies.end(); ++p)
		samples.push_back((*p)->get_name());
	return samples;
}


//////////////////// Progeny ////////////////////

const Progeny *Progeny::copy() const {
	return new Progeny(this->family, this->name, this->mat, this->pat);
}

bool Progeny::with_parents() const {
	return mat != "0" && pat != "0";
}

bool Progeny::is_all_in_samples(const set<string>& samples) const {
	return samples.find(name) != samples.end() &&
		   samples.find(mat) != samples.end() &&
		   samples.find(pat) != samples.end();
}

const Progeny *Progeny::create(const vector<string>& v) {
	assert(v.size() >= 4U);
	return new Progeny(v[0], v[1], v[2], v[3]);
}


//////////////////// PedigreeTable ////////////////////

PedigreeTable::~PedigreeTable() {
	for(auto p = table.begin(); p != table.end(); ++p)
		delete *p;
}

const PedigreeTable *PedigreeTable::filter_with_parents(
									const vector<string>& samples) const {
	vector<const Progeny *>	progs;
	set<string>	set_samples(samples.begin(), samples.end());
	for(auto p = table.begin(); p != table.end(); ++p) {
		if((*p)->with_parents() && (*p)->is_all_in_samples(set_samples))
			progs.push_back((*p)->copy());
	}
	// not destruct Progeny in destructor
	return new PedigreeTable(progs);
}

vector<const Progeny *> PedigreeTable::get_progenies(const string& mat,
													  const string& pat) const {
	vector<const Progeny *>	children;
	set<string>	set_children;	// for avoiding duplication
	pair<string,string>	parents(mat, pat);
	for(auto p = table.begin(); p != table.end(); ++p) {
		if((*p)->parents() == parents) {
			if(set_children.find((*p)->get_name()) == set_children.end()) {
				children.push_back((*p)->copy());
				set_children.insert((*p)->get_name());
			}
		}
	}
	return children;
}

vector<const Family *> PedigreeTable::extract_families() const {
	// collect parents
	set<pair<string,string>>	s;
	for(auto p = this->table.begin(); p != this->table.end(); ++p)
		s.insert((*p)->parents());
	
	vector<const Family *>	families;
	for(auto p = s.begin(); p != s.end(); ++p) {
		const string&	mat = p->first;
		const string&	pat = p->second;
		const auto	progenies = this->get_progenies(mat, pat);
		const Family	*family = new Family(mat, pat, progenies);
		families.push_back(family);
	}
	return families;
}

vector<const Family *> PedigreeTable::make_families(
										const vector<string>& samples) const {
	map<pair<string, string>, vector<const Progeny *>>	progs;
	set<string>	used;
	for(auto p = this->table.begin(); p != this->table.end(); ++p) {
		const Progeny	*prog = *p;
		if(used.find(prog->get_name()) == used.end()) {
			progs[prog->parents()].push_back(prog->copy());
			used.insert(prog->get_name());
		}
	}
	
	vector<const Family *>	families;
	const set<string>	set_samples(samples.begin(), samples.end());
	for(auto p = progs.begin(); p != progs.end(); ++p) {
		const auto	parents = p->first;
		const auto	progenies = p->second;
		// Parent not found in samples of VCF is treated as unknown
		const string	mat = parents.first;
		const string	pat = parents.second;
		if(mat != "0" && pat != "0") {
			const Family	*family = new Family(mat, pat, progenies);
			families.push_back(family);
		}
		else {
			for(auto q = progenies.begin(); q != progenies.end(); ++q) {
				vector<const Progeny *>	new_progs{ *q };
				const Family	*family = new Family(mat, pat, new_progs);
				families.push_back(family);
			}
		}
	}
	std::sort(families.begin(), families.end(),
				[](const Family *lh, const Family *rh)
				{ return lh->parents() < rh->parents(); });
	return families;
}

vector<string> PedigreeTable::check_samples_in_pedigree(
								const vector<string>& samples) const {
	set<string>	ped_samples;
	for(auto p = table.begin(); p != table.end(); ++p)
		ped_samples.insert((*p)->get_name());
	vector<string>	missing_samples;
	for(auto p = samples.begin(); p != samples.end(); ++p) {
		if(ped_samples.find(*p) == ped_samples.end())
			missing_samples.push_back(*p);
	}
	return missing_samples;
}

const PedigreeTable *PedigreeTable::limit_samples(
								const vector<string>& samples) const {
	const auto	missing_samples = check_samples_in_pedigree(samples);
	if(!missing_samples.empty()) {
		throw SamplesException(missing_samples);
	}
	
	set<string>	set_samples(samples.begin(), samples.end());
	vector<const Progeny *>	progs;
	for(auto p = table.begin(); p != table.end(); ++p) {
		if(set_samples.find((*p)->get_name()) != set_samples.end())
			progs.push_back((*p)->copy());
	}
	return new PedigreeTable(progs);
}

vector<string> PedigreeTable::check_parents() const {
	set<string>	set_progs;
	for(auto p = table.begin(); p != table.end(); ++p)
		set_progs.insert((*p)->get_name());
	
	set<string>	missing_parents;
	for(auto p = table.begin(); p != table.end(); ++p) {
		const string&	mat = (*p)->get_mat();
		if(mat != "0" && set_progs.find(mat) == set_progs.end())
			missing_parents.insert(mat);
		const string&	pat = (*p)->get_pat();
		if(pat != "0" && set_progs.find(pat) == set_progs.end())
			missing_parents.insert(pat);
	}
	return vector<string>(missing_parents.begin(), missing_parents.end());
}

vector<vector<string>> PedigreeTable::read_lines(const string& path) {
	ifstream	ifs(path.c_str());
	if(!ifs)
		throw FileNotFoundException(path);
	
	vector<vector<string>>	table;
	vector<string>	not_four_columns_lines;
	string	line;
	while(getline(ifs, line)) {
		if(line.c_str()[line.length()-1] == '\r')
			line = line.substr(0, line.length()-1);
		istringstream	iss(line);
		vector<string>	v;
		string	token;
		while(iss >> token) {
			v.push_back(token);
		}
		if(v.size() != 4) {
			not_four_columns_lines.push_back(line);
		}
		table.push_back(v);
	}
	if(!not_four_columns_lines.empty()) {
		throw FormatException(not_four_columns_lines);
	}
	return table;
}

const vector<const Progeny *> PedigreeTable::remove_duplidated_progenies(
										const vector<const Progeny *>& progs) {
	vector<const Progeny *>	new_progs;
	set<string>	used_names;
	for(auto p = progs.begin(); p != progs.end(); ++p) {
		const string&	prog_name = (*p)->get_name();
		if(used_names.find(prog_name) == used_names.end()) {
			new_progs.push_back(*p);
			used_names.insert(prog_name);
		}
		else {
			cerr << "Progeny " << prog_name << " is duplicated." << endl;
			delete *p;
		}
	}
	return new_progs;
}

const PedigreeTable *PedigreeTable::create(
								const vector<const Progeny *>& progs) {
	const auto	ped = new PedigreeTable(progs);
	const auto	missing_parents = ped->check_parents();
	if(!missing_parents.empty()) {
		delete ped;
		throw ParentsException(missing_parents);
	}
	return ped;
}

const PedigreeTable *PedigreeTable::read(const string& path) {
	const auto	table = PedigreeTable::read_lines(path);
	vector<const Progeny *>	progs;
	for(auto p = table.begin(); p != table.end(); ++p)
		progs.push_back(Progeny::create(*p));
	const auto	unique_progs = remove_duplidated_progenies(progs);
	return create(unique_progs);
}
